Bottom: 047dd99889de634ca308a8f7c5e9053dbc0a343b
Top:    684f57ed391827a478d2c848140dfc2c75cb88c4
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-02-15 13:45:16 +0800

virtio-net: xdp xmit more

Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 9ff959c..5866271 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -84,6 +84,8 @@ struct send_queue {
 
 	/* Name of the send queue: output.$index */
 	char name[40];
+
+	bool xdp_xmit_more;
 };
 
 /* Internal representation of a receive virtqueue */
@@ -404,7 +406,11 @@ static bool virtnet_xdp_xmit(struct virtnet_info *vi,
 		return false;
 	}
 
-	virtqueue_kick(sq->vq);
+	if (sq->vq->num_free < 2)
+		virtqueue_kick(sq->vq);
+	else
+		sq->xdp_xmit_more = true;
+
 	return true;
 }
 
@@ -1015,7 +1021,9 @@ static int virtnet_poll(struct napi_struct *napi, int budget)
 {
 	struct receive_queue *rq =
 		container_of(napi, struct receive_queue, napi);
-	unsigned int r, received;
+	struct virtnet_info *vi = rq->vq->vdev->priv;
+	struct send_queue *sq;
+	unsigned int r, received, qp;
 
 	received = virtnet_receive(rq, budget);
 
@@ -1031,6 +1039,16 @@ static int virtnet_poll(struct napi_struct *napi, int budget)
 		}
 	}
 
+	if (rq->xdp_prog) {
+		qp = vi->curr_queue_pairs - vi->xdp_queue_pairs +
+		     smp_processor_id();
+		sq = &vi->sq[qp];
+		if (sq->xdp_xmit_more) {
+			sq->xdp_xmit_more = false;
+			virtqueue_kick(sq->vq);
+		}
+	}
+
 	return received;
 }
